src/NHibernate/Criterion/GroupedProjection.cs

```diff
@@ -18,13 +17,12 @@ public GroupedProjection(IProjection projection)
 
 		public virtual SqlString ToSqlString(ICriteria criteria, int position, ICriteriaQuery criteriaQuery)
 		{
-			return renderedProjection = projection.ToSqlString(criteria, position, criteriaQuery);
+			return projection.ToSqlString(criteria, position, criteriaQuery);
 		}
 
 		public virtual SqlString ToGroupSqlString(ICriteria criteria, ICriteriaQuery criteriaQuery)
 		{
-			//This is kind of a hack. The hack is based on the fact that ToGroupSqlString always called after ToSqlString.
-			return SqlStringHelper.RemoveAsAliasesFromSql(renderedProjection);
+			return SqlStringHelper.Join(new SqlString(","), CriterionUtil.GetColumnNamesAsSqlStringParts(null, projection, criteriaQuery, criteria));
```

Comment:
This would break.

Reply:
```
3) Failed : NHibernate.Test.NHSpecificTest.NH3609.Fixture.GroupByClauseHasParameterSet
  Expected: No Exception to be thrown
  But was:  <NHibernate.Exceptions.GenericADOException: could not execute query
[ SELECT (case when this_.Name = @p0 then @p1 else @p2 end) as y0_ FROM Entity this_ GROUP BY (case when this_.Name = @p3 then @p4 else @p5 end) ]
  Name:cp0 - Value:  Name:cp1 - Value:1  Name:cp2 - Value:2  Name:cp3 - Value:  Name:cp4 - Value:1  Name:cp5 - Value:2
[SQL: SELECT (case when this_.Name = @p0 then @p1 else @p2 end) as y0_ FROM Entity this_ GROUP BY (case when this_.Name = @p3 then @p4 else @p5 end)] ---> System.Data.SqlClient.SqlException: Column 'Entity.Name' is invalid in the select list because it is not contained in either an aggregate function or the GROUP BY clause.
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlDataReader.TryConsumeMetaData()
   at System.Data.SqlClient.SqlDataReader.get_MetaData()
   at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString)
   at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 timeout, Task& task, Boolean asyncWrite, SqlDataReader ds)
   at System.Data.SqlClient.SqlCommand.ExecuteReader(CommandBehavior behavior)
   at NHibernate.AdoNet.AbstractBatcher.ExecuteReader(DbCommand cmd) in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate/AdoNet/AbstractBatcher.cs:line 226
   at NHibernate.Loader.Loader.GetResultSet(DbCommand st, QueryParameters queryParameters, ISessionImplementor session, IResultTransformer forcedResultTransformer) in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate/Loader/Loader.cs:line 1516
   at NHibernate.Loader.Loader.DoQuery(ISessionImplementor session, QueryParameters queryParameters, Boolean returnProxies, IResultTransformer forcedResultTransformer, QueryCacheResultBuilder queryCacheResultBuilder) in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate/Loader/Loader.cs:line 487
   at NHibernate.Loader.Loader.DoQueryAndInitializeNonLazyCollections(ISessionImplementor session, QueryParameters queryParameters, Boolean returnProxies, IResultTransformer forcedResultTransformer, QueryCacheResultBuilder queryCacheResultBuilder) in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate/Loader/Loader.cs:line 285
   at NHibernate.Loader.Loader.DoList(ISessionImplementor session, QueryParameters queryParameters, IResultTransformer forcedResultTransformer, QueryCacheResultBuilder queryCacheResultBuilder) in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate/Loader/Loader.cs:line 1956
   --- End of inner exception stack trace ---
   at NHibernate.Loader.Loader.DoList(ISessionImplementor session, QueryParameters queryParameters, IResultTransformer forcedResultTransformer, QueryCacheResultBuilder queryCacheResultBuilder) in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate/Loader/Loader.cs:line 1965
   at NHibernate.Impl.SessionImpl.List(CriteriaImpl criteria, IList results) in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate/Impl/SessionImpl.cs:line 1734
   at NHibernate.Impl.CriteriaImpl.List(IList results) in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate/Impl/CriteriaImpl.cs:line 282
   at NHibernate.Impl.CriteriaImpl.List[T]() in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate/Impl/CriteriaImpl.cs:line 294
   at NHibernate.Test.NHSpecificTest.NH3609.Fixture.<>c__DisplayClass4_1.<GroupByClauseHasParameterSet>b__0() in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate.Test/NHSpecificTest/NH3609/Fixture.cs:line 100
--- End of stack trace from previous location where exception was thrown ---
   at NUnit.Framework.Internal.Reflect.DynamicInvokeWithTransparentExceptions(Delegate delegate)
   at NUnit.Framework.Internal.ExceptionHelper.RecordException(Delegate parameterlessDelegate, String parameterName)>
   at NHibernate.Test.NHSpecificTest.NH3609.Fixture.GroupByClauseHasParameterSet() in /home/travis/build/nhibernate/nhibernate-core/src/NHibernate.Test/NHSpecificTest/NH3609/Fixture.cs:line 98
```

Reply:
Yeah - `RemoveAsAliasesFromSql` can't be removed here that easily. I restored it and added logic to handle `IPropertyProjection`.